From 9f3d1c30938cf289b59323336f3dd1ba6d66a699 Mon Sep 17 00:00:00 2001
From: michioxd <michio.haiyaku@gmail.com>
Date: Sun, 25 Jan 2026 20:51:45 +0700
Subject: [PATCH] add missing header

---
 net/dsa/dsa.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/net/dsa/dsa.c b/net/dsa/dsa.c
index 16b19aa..74e25e2 100644
--- a/net/dsa/dsa.c
+++ b/net/dsa/dsa.c
@@ -9,6 +9,7 @@
 
 #include <linux/device.h>
 #include <linux/err.h>
+#include <linux/if_hsr.h>
 #include <linux/list.h>
 #include <linux/module.h>
 #include <linux/netdevice.h>
-- 
2.52.0.windows.1

